apollo-router/src/plugins/telemetry/metrics/apollo.rs

```diff
@@ -326,55 +337,66 @@ impl ApolloMetricsExporter {
     }
 }
 
-fn aggregate(
-    metrics: Vec<Metrics>,
-) -> Vec<(
-    String,
-    ContextualizedStats,
-    HashMap<String, ReferencedFieldsForType>,
-    u64,
-)> {
+fn aggregate(metrics: Vec<Metrics>) -> HashMap<MetricsKey, AggregatedMetrics> {
     let mut aggregated_metrics = HashMap::<_, AggregatedMetrics>::new();
-
     for metric in metrics {
-        *aggregated_metrics
-            .entry(
-                (
-                    metric.client_name.clone(),
-                    metric.client_version.clone(),
-                    metric.stats_report_key.clone(),
-                )
-                    .clone(),
-            )
-            .or_default() += metric;
+        *aggregated_metrics.entry(metric.key.clone()).or_default() += metric;
     }
-
     aggregated_metrics
-        .into_iter()
-        .map(|((client_name, client_version, key), aggregated_metric)| {
-            let mut contextualized_stats = ContextualizedStats {
-                context: Some(StatsContext {
-                    client_name,
-                    client_version,
-                }),
-                ..Default::default()
-            };
-            contextualized_stats.per_type_stat = aggregated_metric
-                .per_type_stat
-                .into_iter()
-                .map(|(k, v)| (k, v.into()))
-                .collect();
-            contextualized_stats.query_latency_stats =
-                Some(aggregated_metric.query_latency_stats.into());
-
-            (
-                key,
-                contextualized_stats,
-                aggregated_metric.referenced_fields_by_type,
-                aggregated_metric.operation_count,
-            )
-        })
-        .collect()
+}
+
+fn to_report(
+    header: ReportHeader,
+    aggregated_metrics: HashMap<MetricsKey, AggregatedMetrics>,
+) -> apollo_spaceport::Report {
+    let time = SystemTime::now()
+        .duration_since(UNIX_EPOCH)
+        .unwrap_or_default();
+    let seconds = time.as_secs();
+    let nanos = time.as_nanos() - (seconds as u128 * 1_000_000_000);
+    let end_time = apollo_spaceport::Timestamp {
+        seconds: seconds as i64,
+        nanos: nanos as i32,
+    };
+
+    let mut report = Report {
+        header: Some(header),
+        traces_per_query: Default::default(),
+        end_time: Some(end_time),
+        operation_count: 0,
+    };
+
+    for (key, metrics) in aggregated_metrics {
+        report.operation_count += metrics.operation_count;
+        if let MetricsKey::Regular {
+            client_name,
+            client_version,
+            stats_report_key,
+        } = key
+        {
+            report.traces_per_query.insert(
+                stats_report_key,
+                apollo_spaceport::TracesAndStats {
+                    trace: vec![],
```

Comment:
fwiw the general model of protobuf is that empty lists are the same as no lists (also 0 is the same as missing) so probably you don't need this (and you definitely don't need internal_traces_contributing_to_stats below which is a debugging feature that we should get rid of anyway, it revealed a bug in Apollo Server and instead of fixing it i just turned off the monitoring because fixing it is hard)

Reply:
Made nicer in [4510527](https://github.com/apollographql/router/pull/1020/commits/4510527600fc3fc5ae074a509d81f0f9a1564cc6), but in practice the code is the same as there is no null in Rust.